karabo_data/geometry2.py

```diff
@@ -205,6 +216,114 @@ def inspect(self, frontview=True):
 
         return ax
 
+    def _tile_dims(self, tileno):
+        """Implement in subclass: which part of module array each tile is.
+        """
+        raise NotImplementedError
+
+
+    @classmethod
+    def from_crystfel_geom(cls, filename):
+        """Read a CrystFEL format (.geom) geometry file.
+
+        Returns a new geometry object.
+        """
+        geom_dict = load_crystfel_geometry(filename)
+        modules = []
+        for p in range(cls.n_modules):
+            tiles = []
+            modules.append(tiles)
+            for a in range(cls.n_asics_per_module):
+                d = geom_dict['panels']['p{}a{}'.format(p, a)]
+                tiles.append(GeometryFragment.from_panel_dict(d))
+        return cls(modules, filename=filename)
+
+    def _get_rigid_groups(self, nquads=4):
+        """Create rigid stings for rigid groups definiton."""
+
+        quads = ','.join(['q{}'.format(q) for q in range(nquads)])
+        modules = ','.join(['p{}'.format(p) for p in range(self.n_modules)])
+
+        prod = product(range(self.n_modules), range(self.n_asics_per_module))
+        rigid_group = ['p{}a{}'.format(p, a) for (p, a) in prod]
+        rigid_string = '\n'
+
+        for nn, rigid_group_q in enumerate(np.array_split(rigid_group, nquads)):
+            rigid_string += 'rigid_group_q{} = {}\n'.format(nn, ','.join(rigid_group_q))
+        rigid_string += '\n'
+        for nn, rigid_group_p in enumerate(np.array_split(rigid_group, self.n_modules)):
+            rigid_string += 'rigid_group_p{} = {}\n'.format(nn, ','.join(rigid_group_p))
+
+        rigid_string += '\n'
+
+        rigid_string += 'rigid_group_collection_quadrants = {}\n'.format(quads)
+        rigid_string += 'rigid_group_collection_asics = {}\n\n'.format(modules)
+        return rigid_string
+
+    def write_crystfel_geom(self, data_path='/entry_1/instrument_1/detector_1/data',
+                            mask_path=None, extra_dim=None, *, filename,
+                            adu_per_ev, clen, photon_energy):
+        """Write this geometry to a CrystFEL format (.geom) geometry file.
+        Parameters
+        ----------
+        data_path  : (str)
+            path to the group that contains the data array in the hdf5 file
+            (default : /entry_1/instrument_1/detector_1/data)
+        mask_path : (str)
+            path to the group that contains the mask array in the hdf5 file
+            (default : /entry_1/instrument_1/detector_1/mask)
+
+        extra_dim (dim_number, index) : collection
+            extra data dimension that has to inserted to read the data. The
+            first number of the tuple refers to the axis number; the second to
+            the index of that axis which holds the data of interest.
```

Comment:
I'm having some trouble understanding this. I guess it's to do with rearranging which axis means what, but the pairs of numbers feel kind of opaque. I'd suggest we leave the data shape hardcoded in this PR (i.e. dim1 = module no, dim 2 & 3 = fast-scan & slow-scan), and deal with making it more flexible as a separate change.

Reply:
well what I meant was the following:
Suppose you have raw data which has the shape of (n_frames, n_modules, 2, fast_scan, slow_scan). If you wanted to get the data you'd access it via the second index of the 3rd dimension. I guess that would be expressed in terms of crystfel geometry code:

```a0p0/dim1 = 0
   a0p0/dim2 = 1
   a0p0/dim3 = fs
   a0p0/dim4 = ss
```
I did this in order to make the hdfsee read the dssc data. In my test this was working fine. To be honest I don't think it's too complicated so I guess we do not have to do this in a seperate PR. Maybe it's just a matter of how it is explained in the doc string

Reply:
OK, I'll have another look tomorrow morning to see if it makes sense to me then, and/or if I can see how to clarify the code at all.

Reply:
What if instead of a parameter to insert an extra dimension, the parameter is a tuple describing all the dimensions, with a default like `('modno', 'ss', 'fs')`? Then to handle an extra dimension for raw data, you would put `dims=('modno', 0, 'fs', 'ss')`.

From a usage perspective, it would be even nicer if it could start from dim0, like `('%', 'modno', 'ss', 'fs')` or `('frame', 'modno', 'ss', 'fs')`. But that might make the implementation trickier, because at the moment `dim0` is described in the header for all panels at once.

Reply:
I guess we could start from `dim0` but require that it's the frame number for now.

`%` is CrystFEL's own placeholder for this in the geometry file, so using that makes it less magic, but obviously something like `frame` is more explicit.

Reply:
Thats a good idea, I'll look into that tmr!

Reply:
Ok I think I am pretty happy with this solution for now.

Reply:
Can you cross check?